File: agent/data_sources/governance.py
"""Live governance data -- Snapshot's public GraphQL API needs no API key."""
from __future__ import annotations
from datetime import datetime, timezone
import httpx

SNAPSHOT_GRAPHQL_URL = "https://hub.snapshot.org/graphql"

SLUG_TO_SNAPSHOT_SPACE = {
    "aave": "aave.eth",
    "lido": "lido-snapshot.eth",
    "uniswap": "uniswapgovernance.eth",
    "compound": "comp-vote.eth",
    "maker": "makerdao.eth",
    "makerdao": "makerdao.eth",
    "curve": "curve.eth",
    "balancer": "balancer.eth",
    "sushiswap": "sushigov.eth",
}

_QUERY = """
query Proposals($space: String!, $first: Int!) {
  proposals(
    first: $first
    skip: 0
    where: { space: $space, state: "closed" }
    orderBy: "created"
    orderDirection: desc
  ) {
    id
    title
    state
    scores
    choices
    end
  }
}
"""


async def fetch_recent_closed_proposals(protocol_slug: str, limit: int = 5) -> tuple[list[dict], str]:
    space = SLUG_TO_SNAPSHOT_SPACE.get(protocol_slug)
    if not space:
        raise ValueError(f"no snapshot space mapped for slug '{protocol_slug}'")

    async with httpx.AsyncClient(timeout=10) as client:
        resp = await client.post(
            SNAPSHOT_GRAPHQL_URL,
            json={"query": _QUERY, "variables": {"space": space, "first": limit}},
        )
        resp.raise_for_status()
        proposals = resp.json()["data"]["proposals"]

    for p in proposals:
        if p["scores"] and p["choices"]:
            winning_index = max(range(len(p["scores"])), key=lambda i: p["scores"][i])
            p["winning_choice"] = p["choices"][winning_index]
        else:
            p["winning_choice"] = None
        p["end_date"] = datetime.fromtimestamp(p["end"], tz=timezone.utc).date().isoformat()

    return proposals, f"{SNAPSHOT_GRAPHQL_URL} (space={space})"

File: agent-ts/packages/core/src/dataSources/governance.ts
/** Live governance data -- Snapshot's public GraphQL API needs no API key. Ported from agent/data_sources/governance.py. */

const SNAPSHOT_GRAPHQL_URL = "https://hub.snapshot.org/graphql";

const SLUG_TO_SNAPSHOT_SPACE: Record<string, string> = {
  aave: "aave.eth",
  lido: "lido-snapshot.eth",
  uniswap: "uniswapgovernance.eth",
  compound: "comp-vote.eth",
  maker: "makerdao.eth",
  makerdao: "makerdao.eth",
  curve: "curve.eth",
  balancer: "balancer.eth",
  sushiswap: "sushigov.eth",
};

const QUERY = `
query Proposals($space: String!, $first: Int!) {
  proposals(
    first: $first
    skip: 0
    where: { space: $space, state: "closed" }
    orderBy: "created"
    orderDirection: desc
  ) {
    id
    title
    state
    scores
    choices
    end
  }
}
`;

export interface GovernanceProposal {
  id: string;
  title: string;
  state: string;
  scores: number[];
  choices: string[];
  end: number;
  winningChoice: string | null;
  endDate: string;
}

export async function fetchRecentClosedProposals(
  protocolSlug: string,
  limit = 5
): Promise<{ proposals: GovernanceProposal[]; source: string }> {
  const space = SLUG_TO_SNAPSHOT_SPACE[protocolSlug];
  if (!space) {
    throw new Error(`no snapshot space mapped for slug '${protocolSlug}'`);
  }

  const resp = await fetch(SNAPSHOT_GRAPHQL_URL, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({ query: QUERY, variables: { space, first: limit } }),
  });
  if (!resp.ok) throw new Error(`Snapshot request failed: ${resp.status}`);
  const json = (await resp.json()) as { data: { proposals: Omit<GovernanceProposal, "winningChoice" | "endDate">[] } };
  const proposals = json.data.proposals;

  const enriched: GovernanceProposal[] = proposals.map((p) => {
    let winningChoice: string | null = null;
    if (p.scores?.length && p.choices?.length) {
      let winningIndex = 0;
      for (let i = 1; i < p.scores.length; i++) {
        if (p.scores[i] > p.scores[winningIndex]) winningIndex = i;
      }
      winningChoice = p.choices[winningIndex];
    }
    const endDate = new Date(p.end * 1000).toISOString().slice(0, 10);
    return { ...p, winningChoice, endDate };
  });

  return { proposals: enriched, source: `${SNAPSHOT_GRAPHQL_URL} (space=${space})` };
}
